### DISCUSSION TITLE
What is the current situation regarding pre-commit?

### DISCUSSION THREAD

**Discussion by TimeTravelPenguin** [prettier/prettier#16227] [General]
I have always used prettier with pre-commit in my Python projects for files such as markdown, XML, and so on.

I just found that [mirrors-prettier](https://github.com/pre-commit/mirrors-prettier) has recently been archived, even though the [prettier website](https://prettier.io/docs/en/precommit.html) still suggests it as an option.

Is there an alternative, workaround, or other means to keep my projects functioning?
What versions of prettier are still compatible? And lastly, are there currently any replacements currently in development?

**Comment by paddyroddy** [comment]
I hadn't spotted that @TimeTravelPenguin, thanks for identifying. I too always use `prettier` through `pre-commit`.

**Reply by paddyroddy** [reply]
From the README

> prettier made some changes that breaks plugins entirely

I suspect this is the maintainer of `pre-commit` who has made this decision

**Comment by datalogics-kam** [comment]
I was promulgating `prettier` to another project at work, and went to see if there was a new version and saw that it was archived. An official statement of direction for fixing `pre-commit` issues or continued support of `pre-commit` would be welcome; I don't even mind changing the repo that `pre-commit` uses.

> prettier made some changes that breaks plugins entirely

Private-life-me opened https://github.com/prettier/prettier/issues/15696, but my intent certainly was not to get Prettier to be delisted from pre-commit.

Prettier is also not listed in the [supported hooks](https://pre-commit.com/hooks.html), and I certainly don't consider `clang-format`(❗ ) a viable alternative.

`mirrors-prettier` is not the only [archived pre-commit repo](https://github.com/pre-commit?q=&type=archived&language=&sort=), and I might suggest that the Prettier team take ownership by perhaps reactivating https://github.com/prettier/pre-commit.

Prettier is just too useful, and it's especially useful in pre-commit. Suggestions on how to use it from the shell with pre-commit would be useful, as would an official Docker image.

**Comment by trallnag** [comment]
[Here](https://github.com/nsidc/earthaccess/issues/555#issuecomment-2241365784) someone pointed out a fork of the archived mirror https://github.com/rbubley/mirrors-prettier

Maybe the prettier org can adopt the mirror?

**Comment by Crystal-RainSlide** [comment]
There isn't much infomation about this situation on https://pre-commit.com/ or any of its README.

Also, there isn't much original issue links from asottile, who archived pre-commit's prettier plugin.  I fear many of them are unavailable after the https://github.com/pre-commit/mirrors-prettier repo is archived.  His idea about this might also be lost, and buried under all those "duplicate, try search harder next time" replies.

But anyway, here is some of the issues about version / releases of the pre-commit mirrors, including the faulty `mirrors-prettier`:
- Unavailable on GitHub due to archived repo, but **PLEASE READ THIS FIRST**:
  https://web.archive.org/web/20230315204029/https://github.com/pre-commit/mirrors-prettier/issues/29

- https://github.com/pre-commit/mirrors-prettier/pull/39
  > Anton Agestam: [Note that I have no high hopes of a reasonable response to this PR.](https://github.com/pre-commit/mirrors-prettier/issues/29#issuecomment-1331093710)
- Feb 22, 2022, closed: https://github.com/pre-commit/pre-commit-mirror-maker/issues/123
  > asottile: the history only moves forward, backporting is intentionally not supported
- Sep 20, 2022 ~ Dec 5, 2023, closed & locked: https://github.com/pre-commit/pre-commit-mirror-maker/issues/146
  > asottile: there's a bunch of duplicates in pre-commit/pre-commit -- tags are just strings and not parsed for versions (there's no agreed upon standard versioning or naming scheme for tags and I don't want to get into parsing them).
  >
  > if you don't want to autoupdate don't run autoupdate -- or use the --repo to select the repos you want to update -- or git checkout -p the repos you don't want

<details>
<summary>Duplicates</summary>

- Sep 4, 2023, duplicate, closed & locked: https://github.com/pre-commit/pre-commit/issues/2988
- Dec 20, 2023, duplicate, closed: https://github.com/pre-commit/pre-commit-mirror-maker/issues/202
- Jan 9, 2024, duplicate, closed: https://github.com/pre-commit/pre-commit/issues/3108
- Jan 17, 2024, (duplicate), closed & locked: https://github.com/pre-commit/pre-commit-mirror-maker/issues/206
- Feb 5, 2024, duplicate, closed & locked: https://github.com/pre-commit/pre-commit-mirror-maker/issues/207
- Feb 15, 2024, duplicate, closed & locked: https://github.com/pre-commit/pre-commit/issues/3128
</details>

- Feb 23, 2024 ~ Feb 24, 2024, closed & locked: https://github.com/pre-commit/pre-commit/issues/3133
- Apr 11, 2024 ~ Apr 12, 2024, closed & locked: https://github.com/pre-commit/pre-commit-mirror-maker/issues/213
  > asottile: no I don't have a solution as you can see. noone has suggested anything that works just hacks that get what they want without regard for compatibility or safety
  >
  > frankly I'm just going to archive mirrors-prettier since it's not worth my trouble

**Comment by JoC0de** [comment]
I created a pre-commit mirror that solves the issue with not working pre-commit plugins in version 3 and later.
It also skips alpha versions.
https://github.com/JoC0de/pre-commit-prettier

**Comment by v-goncharenko** [comment]
I used to use prettier for python projects as well. Why not to add support for pre-commit tool to this repo directly?

**Comment by Goldziher** [comment]
Maybe target `prek`? its the new pre-commit rewrite in rust, and the maintainer is easier to work with than the pre-commit maintainer.

**Comment by 34j** [comment]
I added non-monotonic version increment support to pre-commit-mirror-maker. It only requried **4 lines** of modification and 1 file deletion, as you may see in https://github.com/pre-commit/pre-commit-mirror-maker/pull/261/changes: https://github.com/mirrors-prettier/pre-commit-mirror-maker-non-monotonic

It is used in: https://github.com/mirrors-prettier/mirrors-prettier

```yml
-   repo: https://github.com/mirrors-prettier/mirrors-prettier
    rev: ''  # Use the sha / tag you want to point at
    hooks:
    -   id: prettier
```

**Comment by Interpause** [comment]
One more solution for those who are mindful of pulling in additional dependencies is to manage the hook yourself using the `repo: local` escape hatch (also works with prek):

```yaml
  - repo: local
    hooks:
      - id: prettier
        name: prettier
        entry: prettier
        args: [--check, --ignore-unknown]
        language: node
        types: [text]
        additional_dependencies:
          - prettier@3.9.5
```

**Reply by sethrj** [reply]
Awesome! Perfect solution.

**Comment by Maher-Reven** [comment]
A status check, since this thread spans two years and the original premise has moved on.

**The docs no longer point at the archived mirror.** [`docs/precommit.md`](https://github.com/prettier/prettier/blob/main/docs/precommit.md) now lists lint-staged, pretty-quick, Husky.Net, git-format-staged, Lefthook (added in #19191) and a plain shell script. So the specific problem @TimeTravelPenguin opened this with is gone.

**What is still missing is any section for `pre-commit` itself.** That is what most people in this thread actually use, and the answer is currently spread across eight comments here rather than in the docs.

@Interpause's `repo: local` recipe is the one that works today without depending on any third-party mirror. I ran it rather than just repeating it:

```yaml
# .pre-commit-config.yaml
repos:
  - repo: local
    hooks:
      - id: prettier
        name: prettier
        entry: prettier
        args: [--write, --ignore-unknown]
        language: node
        types: [text]
        additional_dependencies:
          - prettier@3.9.6
```

What I checked, on a scratch repo with `pre-commit run --all-files`:

- It formats. A messy `.js` and a messy `.md` both came out correctly formatted.
- It is idempotent. The second run reports `Passed`, so it does not fight your commits.
- `.prettierrc` is respected — `{ "semi": false, "singleQuote": true }` was applied.
- `.prettierignore` is respected — a listed file was skipped entirely.

Two things worth knowing before adopting it:

- `language: node` makes `pre-commit` install its **own** Node toolchain and its own copy of Prettier in an isolated environment. That is the point — it works in a Python project with no `node_modules` — but it also means the version in `additional_dependencies` is pinned separately from your `package.json` and will silently drift. If your repo also has Prettier as a devDependency, add a Renovate/Dependabot rule for `.pre-commit-config.yaml` or the two will diverge.
- Use `args: [--check, --ignore-unknown]` instead if you want the hook to fail rather than rewrite files, which is usually what you want in CI.

[`prek`](https://github.com/j178/prek) takes the same config unchanged, for anyone who followed @Goldziher's suggestion.

Question for the maintainers: would a short `pre-commit` section in `docs/precommit.md` be welcome, in the same shape as the Lefthook one from #19191? Happy to open a PR with the above if so, but I would rather ask than spend a PR you do not want.

**Reply by sethrj** [reply]
Along those lines, since `node` is supported by pre-commit (and perhaps it wasn't when this thread started?), maybe it's possible to just add a pre-commit-config.yaml that invokes `bin/prettier`?

**Reply by Maher-Reven** [reply]
@sethrj Yes — and I tested it rather than guessing, because I expected it to fail and it doesn't.

One naming correction first: a repo that *provides* a hook ships **`.pre-commit-hooks.yaml`**; `.pre-commit-config.yaml` is the file a consumer writes. So this would be a new file at the root of `prettier/prettier`:

```yaml
# .pre-commit-hooks.yaml
- id: prettier
  name: prettier
  description: Format files with Prettier.
  entry: prettier
  language: node
  types: [text]
  args: [--write, --ignore-unknown]
```

and then anyone could write, with no mirror in the middle:

```yaml
repos:
  - repo: https://github.com/prettier/prettier
    rev: 3.9.6
    hooks:
      - id: prettier
```

`node` support isn't new, but two things about this repo made me doubt it would work: `bin/prettier.cjs` runs from `src/`, which only exists in the source tree and not in the published package, and the repo's own `package.json` is itself named `prettier` *and* carries `"prettier": "3.9.6"` in `devDependencies` for self-formatting. I expected the global install to collide on the name and quietly hand you the published copy instead of the checkout.

It doesn't. To be sure of the provenance I used a fingerprint rather than a version string: I put the manifest above on a branch that also carries my open fix for #19955 (PR #20075), pointed `rev` at that exact SHA, and fed the hook the input from that bug.

```
input   x2 = (a) => (a /* ! */ // K
        )

output  x2 = (a) => a; /* ! */ // K
```

Published 3.9.6 prints that as two lines with the comment in the wrong place — that is the bug. Getting the one-line form back is only possible if the hook ran the source at the pinned rev. So `rev` really does pin the formatter here, which is strictly better than a mirror, where `rev` pins the mirror and the version is a second thing to keep in sync.

The cost is the honest part. `language: node` installs its own Node, then `npm install`s the repo's **source** dependency tree — 90 runtime deps for the build, not the zero-dependency published package:

| | |
|:--|:--|
| cold run (clone + install + format) | ~14 s, on an already-warm `~/.npm` |
| warm run | ~0.4 s |
| `PRE_COMMIT_HOME` footprint | ~400 MB, of which ~333 MB is the node env |

A mirror installs the published tarball and costs a few MB. So this is a real trade: authoritative versioning and no third-party repo, against a much heavier environment on every contributor's machine and every CI runner that doesn't cache `~/.cache/pre-commit`. I'd rather lay that out than argue for one.

Two practical notes if anyone picks this up:

- It only helps from the *next* tag onwards. `rev` checks out the manifest at that ref, so `rev: 3.9.6` cannot work until a release exists that contains the file.
- I tested against a branch of my own fork, since `prettier/prettier` ships no manifest today. Same mechanism, same `repo:`/`rev:` resolution — but that is a fork, not this repo, so treat the numbers as indicative.

This doesn't change what I'd suggest for the docs. `repo: local` still needs nothing from the maintainers and works on every released version, so it's the right thing to document now; a shipped manifest would be a separate, larger decision about owning a distribution channel.
